viewer/packages/cad-model/src/batching/MultiBufferBatchingManager.ts

```diff
@@ -0,0 +1,313 @@
+      const treeIndex = treeIndexAttribute.getX(i);
+      decrementOrDeleteIndex(mesh.userData.treeIndices, treeIndex);
+    }
+  }
+
+  private processGeometries(parsedGeometry: Required<ParsedGeometry>, sectorBatch: SectorBatch) {
+    const instanceBatch = this.getOrCreateInstanceBatch(parsedGeometry);
+    this.batchInstanceAttributes(parsedGeometry.geometryBuffer, parsedGeometry.instanceId, instanceBatch, sectorBatch);
+  }
+
+  private batchInstanceAttributes(
+    bufferGeometry: BufferGeometry,
+    instanceId: string,
+    instanceBatch: InstanceBatch,
+    sectorBatch: SectorBatch
+  ) {
+    const interleavedBufferView = GeometryBufferUtils.getInstanceAttributesSharedView(bufferGeometry);
+    const batchBuffer = this.getBatchBufferToFill(instanceBatch);
+
+    const length = interleavedBufferView.byteLength;
+    const offset = interleavedBufferView.byteOffset;
+    const interleavedAttributesView = new Uint8Array(interleavedBufferView.buffer, offset, length);
+
+    const { batchId, bufferIsReallocated, updateRange } = batchBuffer.buffer.add(interleavedAttributesView);
+
+    const sectorInstanceData: SectorInstanceData = { batchBuffer, batchId, instanceCount: 0 };
+
+    sectorBatch.instanceBatches.set(instanceId, sectorInstanceData);
+
+    GeometryBufferUtils.getAttributes(batchBuffer.mesh.geometry, InterleavedBufferAttribute).forEach(
+      ({ attribute }) => {
+        attribute.data.needsUpdate = true;
+      }
+    );
+
+    const sourceInstanceAttributes = GeometryBufferUtils.getAttributes(bufferGeometry, InterleavedBufferAttribute);
+
+    const treeIndexInterleavedAttribute = this.getTreeIndexAttribute(sourceInstanceAttributes);
+
+    for (let i = 0; i < treeIndexInterleavedAttribute.count; i++) {
+      incrementOrInsertIndex(batchBuffer.mesh.userData.treeIndices, treeIndexInterleavedAttribute.getX(i));
+    }
+
+    if (bufferIsReallocated) {
+      this.reallocateBufferGeometry(batchBuffer);
+    } else {
+      this.updateInstanceAttributes(sourceInstanceAttributes, updateRange);
+    }
+
+    const instanceCount = sourceInstanceAttributes[0].attribute.count;
+    sectorInstanceData.instanceCount = instanceCount;
+
+    batchBuffer.mesh.visible = true;
+    batchBuffer.mesh.count += instanceCount;
+  }
+
+  private reallocateBufferGeometry({ buffer, mesh }: BatchBuffer) {
+    const defragmentedBufferGeometry = this.createDefragmentedBufferGeometry(mesh.geometry, buffer);
+    mesh.geometry.dispose();
+    mesh.geometry = defragmentedBufferGeometry;
+  }
+
+  private updateInstanceAttributes(
+    instanceAttributes: {
+      name: string;
+      attribute: InterleavedBufferAttribute;
+    }[],
+    updateRange: {
+      byteOffset: number;
+      byteCount: number;
+    }
+  ) {
+    instanceAttributes.forEach(namedAttribute => {
+      const attribute = namedAttribute.attribute;
+      this.extendUpdateRange(attribute, updateRange);
+      attribute.data.needsUpdate = true;
+    });
+  }
```

Comment:
Just to be sure - do all the `namedAttribute.attribute`'s share the same buffer? Do we need to mark all of the attributes for update?

Reply:
They have separate InterleavedBuffer instances, but they point to the same underlying data.